File: src/RenderEngine.Core/Point3D.cs
namespace RenderEngine.Core
{
    public class Point3D : Primitive3D
    {
        public Point3D(float x, float y, float z)
            : base(x, y, z, 1)
        { }

        internal Point3D(Primitive3D primitive)
            : base(primitive)
        { }

        public static Vector3D operator -(Point3D left, Point3D right)
        {
            return left.Subtract(right).ToVector3D();
        }

        public static Point3D operator +(Point3D left, Vector3D right)
        {
            return left.Add(right).ToPoint3D();
        }

        public static Point3D operator -(Point3D left, Vector3D right)
        {
            return left.Subtract(right).ToPoint3D();
        }
    }
}

File: lib/ray-tracer-core/Point3D.cpp
#include "Mathf.h"
#include "Primitive3D.h"
#include "Vector3D.h"
#include "Point3D.h"

Vector3D Point3D::operator-(const Point3D& other)
{
	return Vector3D(X() - other.X(), Y() - other.Y(), Z() - other.Z());
}

Point3D Point3D::operator+(const Vector3D& other)
{
	return Point3D(X() + other.X(), Y() + other.Y(), Z() + other.Z());
}

Point3D Point3D::operator-(const Vector3D& other)
{
	return Point3D(X() - other.X(), Y() - other.Y(), Z() - other.Z());
}

ostream& operator<<(ostream& out, const Point3D& p)
{
	out << "Point3D";
	out << (Primitive3D<double>)p;

	return out;
}